Node affinity scheduling strategy (#23381)
ray-project/ray: doc/source/ray-core/tasks/scheduling.rst

Scheduling Strategy
-------------------
Tasks support ``scheduling_strategy`` option to specify the strategy used to decide the best node among available nodes.
Currently the supported strategies are ``"DEFAULT"``, ``"SPREAD"`` and

> ``ray.util.scheduling_strategies.NodeSchedulingStrategy(node_id, soft)``.

Reviewer: ```suggestion
``ray.util.scheduling_strategies.NodeSchedulingStrategy(node_id, soft: bool)``.
```

"DEFAULT" is the default strategy used by Ray. With the current implementation, Ray will try to pack tasks on nodes
until the resource utilization is beyond a certain threshold and spread tasks afterwards.

"SPREAD" strategy will try to spread the tasks among available nodes.

> NodeSchedulingStrategy is the low level strategy that allows a task to be scheduled onto a particular node specified by the hex node id.

Reviewer 2: (apply same textual fixes from above here).

> NodeSchedulingStrategy is the low level strategy that allows a task to be scheduled onto a particular node specified by the hex node id.

Reviewer: ```suggestion
NodeSchedulingStrategy is a low-level strategy that allows a task to be scheduled onto a particular node, specified by `ray.NodeID`.
```

The ``soft`` flag specifies whether the task is allowed to run somewhere else if the specified node doesn't exist (e.g. the node dies)

> or is infeasible for the required resources. If it is True, the task will be scheduled onto other feasible nodes under such case.

Reviewer: ```suggestion
or is infeasible because it does not have the resources required to run the task. In these cases, if ``soft`` is True, the task will be scheduled onto a different feasible node.
```

> If it's False, the task will fail with ``TaskUnschedulableError``.

Reviewer: ```suggestion
Otherwise, the task will fail with ``TaskUnschedulableError``.
```

As long as the specified node is alive and feasible, the task will only run there
regardless of the ``soft`` flag. This means if the node currently has no available resources, the task will wait until resources
become available.

> This strategy should ONLY be used if other high level scheduling strategies (e.g. placement group) cannot give the

Reviewer: ```suggestion
This strategy should *only* be used if other high level scheduling strategies (e.g. placement group) cannot give the
```

Also link to placement group page?

desired task placements. It has the following known limitations:
1. It's a low level strategy which prevents optimizations by a smart scheduler.
2. It cannot fully utilize an autoscaling cluster since node ids must be known when the tasks are created.

> 3. Users normally don't have enough information to make the best static task placement decision
> especially in a multi-tenant cluster: for example, they don't know what else are being scheduled onto those nodes as well.

Reviewer: ```suggestion
3. It can be difficult to make the best static task placement decision
especially in a multi-tenant cluster: for example, an application won't know what else is being scheduled onto the same nodes.
```

        # Spread tasks across the cluster.
        [spread_function.remote() for i in range(100)]

        @ray.remote
        def node_function():
            return 3

        # Only run the task on the local node.
        node_function.options(
            scheduling_strategy=NodeSchedulingStrategy(
                node_id = ray.get_runtime_context().node_id.hex(),
                soft = False,
            )
        ).remote()

        # Run the task on the remote node if possible.
        node_function.options(

>             scheduling_strategy=NodeSchedulingStrategy(

Reviewer 3: on naming: will `NodeAffinityStrategy` be a better name?

Author: Yea, sounds a better name! Should it be `NodeAffinitySchedulingStrategy` or it's too long.

Reviewer 3: yeah `NodeAffinityStrategy` seems having the right balance.

Author: @stephanie-wang @ericl wdyt?

                node_id = remote_node_id,
                soft = True,
            )

>         ).remote()

Reviewer: One more thing for this doc: can we link to the docs for APIs that return node IDs that the user can pass? Specifically the get_runtime_context call and the call to get all of the node IDs in the current cluster.

Reviewer: Also, is this code runnable? I didn't see `remote_node_id` defined. We should make it runnable if possible!
